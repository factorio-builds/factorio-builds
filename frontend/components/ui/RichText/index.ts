export { default } from "./rich-text.component"
