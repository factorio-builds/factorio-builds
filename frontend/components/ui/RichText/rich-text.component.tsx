import React from "react"
import ItemIcon from "../ItemIcon"
import * as S from "./rich-text.styles"
import useParseRichText, { IParsedRichTextNode } from "./useParseRichText.hook"

interface IRichText {
  input: string
  prefix?: JSX.Element
}

interface IRichTextNodeProps {
  node: IParsedRichTextNode
  index: number
}

function RichTextNode({ node, index }: IRichTextNodeProps): JSX.Element {
  if (node.type === "text") {
    return <span key={index}>{node.value}</span>
  }

  if (node.type === "item") {
    return <ItemIcon key={index} type="item" name={node.value} />
  }

  return (
    <span style={{ color: node.value }}>
      {node.children.map((child, i) => {
        return <RichTextNode key={i} index={i} node={child} />
      })}
    </span>
  )
}

function RichText(props: IRichText): JSX.Element {
  const parts = useParseRichText(props.input)

  const formatted = React.useMemo(() => {
    if (!props.input) {
      return "[unnamed]"
    }

    return parts.map((part, index) => {
      return <RichTextNode key={index} node={part} index={index} />
    })
  }, [props.input])

  return (
    <S.RichTextWrapper>
      {props.prefix} {formatted}
    </S.RichTextWrapper>
  )
}

export default RichText
