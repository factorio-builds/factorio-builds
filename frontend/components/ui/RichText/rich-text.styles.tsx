import { styled } from "../../../design/stitches.config"

export const RichTextWrapper = styled("span", {
  display: "inline-block",

  "&::after": {
    content: "",
    clear: "both",
    display: "table",
  },

  "> :first-child": {
    marginLeft: 0,
  },

  "> :last-child": {
    marginRight: 0,
  },

  img: {
    float: "left",
    width: "auto",
    height: "1em",
    margin: "0 0.25em",
  },
})
