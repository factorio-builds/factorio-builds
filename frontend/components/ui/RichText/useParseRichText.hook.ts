import { useMemo } from "react"
import { IParsedRichTextNode, RichTextParser } from "../../../utils/rich-text"

function useParseRichText(text?: string): IParsedRichTextNode[] {
  return useMemo(() => {
    if (!text) {
      return []
    }

    const parsed = RichTextParser.value.parse(text)
    if (parsed.status) {
      return parsed.value
    } else {
      console.log("===")
      console.log(text)
      console.log(parsed)
    }
    return [{ type: "text", value: text }]
  }, [text])
}

export default useParseRichText
