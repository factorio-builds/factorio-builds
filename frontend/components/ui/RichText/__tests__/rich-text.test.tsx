import React from "react"
import { render } from "@testing-library/react"
import RichText from "../index"

describe("<RichText />", () => {
  it("parses a title with no icons", () => {
    const { getByText } = render(<RichText input="build name" />)

    expect(getByText("build name")).toBeTruthy()
  })

  it("returns a default title when passed an empty string", () => {
    const { getByText } = render(<RichText input="" />)

    expect(getByText("[unnamed]")).toBeTruthy()
  })

  it("renders a prefix", () => {
    const { getByText } = render(
      <RichText prefix={<div>prefix</div>} input="build name" />
    )

    expect(getByText("build name")).toBeTruthy()
    expect(getByText("prefix")).toBeTruthy()
  })

  it("parses a title with item icon", () => {
    const { container, getByText, getAllByTestId } = render(
      <RichText input="[item=transport-belt] build name" />
    )

    const icons = getAllByTestId("item-icon") as HTMLImageElement[]

    expect(icons).toHaveLength(1)
    expect(icons[0].src).toBe(
      "https://api.local.factorio.tech/assets/icon/64/item/transport-belt.png"
    )
    expect(getByText("build name")).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it("parses a title with multiple item icons", () => {
    const { container, getByText, getAllByTestId } = render(
      <RichText input="[item=transport-belt] build name [item=fast-transport-belt]" />
    )

    const icons = getAllByTestId("item-icon") as HTMLImageElement[]

    expect(icons).toHaveLength(2)
    expect(icons[0].src).toBe(
      "https://api.local.factorio.tech/assets/icon/64/item/transport-belt.png"
    )
    expect(icons[1].src).toBe(
      "https://api.local.factorio.tech/assets/icon/64/item/fast-transport-belt.png"
    )
    expect(getByText("build name")).toBeTruthy()
    expect(container).toMatchSnapshot()
  })
})
