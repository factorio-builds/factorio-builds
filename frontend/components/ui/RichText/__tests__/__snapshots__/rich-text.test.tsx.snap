// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RichText /> parses a title with item icon 1`] = `
<div>
  <span
    class="rich-textstyles__RichTextWrapper-sc-1hmb2ya-0 huDzTU"
  >
     
    <img
      class="item-iconstyles__ItemIconWrapper-sc-16iffik-0 llnkpE"
      data-testid="item-icon"
      name="transport-belt"
      src="https://api.local.factorio.tech/assets/icon/64/item/transport-belt.png"
      type="item"
    />
    <span>
       build name
    </span>
  </span>
</div>
`;

exports[`<RichText /> parses a title with multiple item icons 1`] = `
<div>
  <span
    class="rich-textstyles__RichTextWrapper-sc-1hmb2ya-0 huDzTU"
  >
     
    <img
      class="item-iconstyles__ItemIconWrapper-sc-16iffik-0 llnkpE"
      data-testid="item-icon"
      name="transport-belt"
      src="https://api.local.factorio.tech/assets/icon/64/item/transport-belt.png"
      type="item"
    />
    <span>
       build name 
    </span>
    <img
      class="item-iconstyles__ItemIconWrapper-sc-16iffik-0 llnkpE"
      data-testid="item-icon"
      name="fast-transport-belt"
      src="https://api.local.factorio.tech/assets/icon/64/item/fast-transport-belt.png"
      type="item"
    />
  </span>
</div>
`;
